### ISSUE TITLE
Setting -backend=false in terraform init

### ISSUE THREAD

**Issue by metmajer** [newcontext-oss/kitchen-terraform#268]
We are working on integrating Consul as a Terraform backend to centrally store Terraform state. Upon invoking `bundle kitchen test`, Terraform bails out with the following error:

```
$$$$$$ Running command `terraform init -input=false -lock=true -lock-timeout=0s  -force-copy -backend=true  -get=true -get-plugins=true  -verify-plugins=true /Users/metmajer/Workspace/my-awesome-stack/test/fixtures/default`
       Initializing modules...
       - module.my-awesome-stack

       Initializing provider plugins...

       Terraform has been successfully initialized!
$$$$$$ Running command `terraform workspace select kitchen-terraform-default-aws`
       Backend reinitialization required. Please run "terraform init".
       Reason: Initial configuration of the requested backend "consul"

       The "backend" is the interface that Terraform uses to store state,
       perform operations, etc. If this message is showing up, it means that the
       Terraform configuration you're using is using a custom configuration for
       the Terraform backend.

       Changes to backend configurations require reinitialization. This allows
       Terraform to setup the new configuration, copy existing state, etc. This is
       only done during "terraform init". Please run that command now then try again.

       If the change reason above is incorrect, please verify your configuration
       hasn't changed and try again. At this point, no changes to your existing
       configuration or state have been made.

       Failed to load backend: Initialization required. Please see the error message above.
```

The error occurs, because, due to `terraform init -backend=true`, Terraform expects a backend configuration, which is partial in `backend.tf` at this point in time (i.e. the necessary parts to complete the backend configuration are not present in testing).

I question the need for a remote, locking backend in testing, especially when testing occurs in development and when you are working on a bunch of Terraform modules which all have the same setup in their respective `kitchen.yml`. Since they're all configured equally, they'll conflict when tests run concurrently (as multiple modules share the same workspace name):

Would it be a good idea to allow `-backend=false`? Cheers!

**Comment by ncs-alane** [comment]
Hi @metmajer.

Thank you for your interest in the project!

There is no requirement to use a remote, locking backend in testing; that decision ultimately lies with the Terraform module which you are testing. If the module under test has a backend then it must be configured before Terraform can operate. Setting `terraform init -backend=false` would not solve this problem. From [Command: init - Backend Initialization](https://www.terraform.io/docs/commands/init.html#backend-initialization):

> To skip backend configuration, use -backend=false. Note that some other init steps require an initialized backend, so it is recommended to use this flag only when the working directory was already previously initialized for a particular backend.

I believe that your options are to use the driver plugin's `backend_configurations` attribute to configure the backend as appropriate, or redesign the module under test so that it does not enforce a remote backend. As for solving the conflict of namespaces during tests, my general recommendation is to design modules under test with a name variable which can be populated with the name of the testing user or a random string.

**Comment by metmajer** [comment]
Hi @ncs-alane, thanks for pointing this out to me!

The idea was to disable the use of a remote backend in testing and refer to a local backend instead. I had assumed that setting `-backend=false` through kitchen-terraform would enforce the intended behavior.

Following your comments, I now have:
- a `backend.tf` using a `consul` backend type
- a `backend.tf` using a `local` backend type in `test/fixtures/default` (my `root_module_directory`)

Which yields:
```
-----> Starting Kitchen (v1.22.0)
$$$$$$ Running command `terraform version`
       Terraform v0.11.7
       + provider.aws v1.28.0
       + provider.random v1.3.1

       Your version of Terraform is out of date! The latest version
       is 0.11.8. You can update by downloading from www.terraform.io/downloads.html
$$$$$$ Terraform v0.11.7 is supported
-----> Destroying <default-aws>...
$$$$$$ Running command `terraform init -input=false -lock=true -lock-timeout=0s  -force-copy -backend=true  -get=true -get-plugins=true  -verify-plugins=true /Users/metmajer/Workspace/my-awesome-stack/test/fixtures/default`
       Initializing modules...
       - module.my-awesome-stack

       Initializing the backend...

       Initializing provider plugins...

       Terraform has been successfully initialized!
$$$$$$ Running command `terraform workspace new kitchen-terraform-default-aws`
       Backend reinitialization required. Please run "terraform init".
       Reason: Backend configuration changed for "consul"

       The "backend" is the interface that Terraform uses to store state,
       perform operations, etc. If this message is showing up, it means that the
       Terraform configuration you're using is using a custom configuration for
       the Terraform backend.

       Changes to backend configurations require reinitialization. This allows
       Terraform to setup the new configuration, copy existing state, etc. This is
       only done during "terraform init". Please run that command now then try again.

       If the change reason above is incorrect, please verify your configuration
       hasn't changed and try again. At this point, no changes to your existing
       configuration or state have been made.

       Failed to load backend: Initialization required. Please see the error message above.
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Destroy failed on instance <default-aws>.  Please see .kitchen/logs/default-aws.log for more details
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration

make: *** [test] Error 20
```

The reason is obviously that `terraform init` is invoked in the top level directory, not in the `root_module`, hence the `local` backend is ignored. As an alternative, I would like to use `backend_configurations`, but its use requires on a partial backend configuration.

Currently, I just can't see a way to use a `local` backend in testing only. Any ideas? Thanks!

**Comment by ncs-alane** [comment]
Based on that output, it appears that you are running an older version of Kitchen-Terraform, as Kitchen-Terraform 4.0.0 sets the working directory to `root_module_directory`. Can you confirm if this is true?

**Comment by metmajer** [comment]
That's true! Thanks for helping out. I tried to upgrade, but ran into dependency issues in #269.

**Comment by metmajer** [comment]
The issue with the working directory has resolved with 4.0.0. Cheers!
